pyanaconda/modules/network/nm_client.py

```diff
@@ -557,37 +594,62 @@ def add_connection_from_ksdata(nm_client, network_data, device_name, activate=Fa
 def add_connection_sync(nm_client, connection):
     """Add a connection synchronously and optionally activate asynchronously.
 
+    Synchronous run is implemented by running a blocking GMainLoop with
+    GMainContext belonging to the nm_client created for the calling Task.
+
+    :param nm_client: NetoworkManager client
+    :type nm_client: NM.NMClient
     :param connection: connection to be added
     :type connection: NM.SimpleConnection
     :return: added connection or None on timeout
     :rtype: NM.RemoteConnection
     """
-    sync_queue = Queue()
-
-    def finish_callback(nm_client, result, sync_queue):
-        con, result = nm_client.add_connection2_finish(result)
-        log.debug("connection %s added:\n%s", con.get_uuid(),
-                  con.to_dbus(NM.ConnectionSerializationFlags.NO_SECRETS))
-        sync_queue.put(con)
-
-    nm_client.add_connection2(
-        connection.to_dbus(NM.ConnectionSerializationFlags.ALL),
-        (NM.SettingsAddConnection2Flags.TO_DISK |
-         NM.SettingsAddConnection2Flags.BLOCK_AUTOCONNECT),
-        None,
-        False,
-        None,
-        finish_callback,
-        sync_queue
-    )
+    info = "NMClient.add_connection2"
 
-    try:
-        ret = sync_queue.get(timeout=CONNECTION_ADDING_TIMEOUT)
-    except Empty:
-        log.error("Adding of connection %s timed out.", connection.get_uuid())
-        ret = None
+    with MainLoopRun(
+        info,
+        nm_client.get_main_context(),
+        timeout=CONNECTION_ADDING_TIMEOUT
+    ) as run:
+
+        def _finish_callback(nm_client, async_result):
```

Comment:
I can't help but the version with context manager seems somehow more readable to me.
Also, as I said, the reslut.recieved data would get only res in case of connection.update2

Reply:
Ok, let's look at it next week and figure out what we can do with the context manager to make it easier to use and understand.

Reply:
Your solution is better. I've updated the PR adding it (slightly modified).
And overall with your other suggestions the patch now looks better (and ready for eventual review by Thomas).